### DISCUSSION TITLE
accessing dataSetRow["typeCol"] inside a chart script

### DISCUSSION THREAD

**Discussion by arass** [eclipse-birt/birt#1848] [Q&A]
Have a running report which displays dataSetRow["typeCol"] as a label just fine.
Inside the report there are charts that work fine. 
But i need to know what the dataSetRow["typeCol"]  value is (inside some script behind a chart? ) so that I can set max scope based on dataSetRow["typeCol"] 

if I try and grab dataSet or series.getData() - that only contains the data for the current axis and none of the other columns (cause of grouping, I guess). 

Thanks in advance.

**Comment by hvbtup** [comment]
In never used BIRT charts myself, but the canonical way to access values in scripts for layout items is `row["..."`], not `dataSetRow["..."]`. I think one should use the latter only inside the definition of a binding.

**Comment by arass** [comment]
Methods are different from regular script. Bunch of stuff d/n work right. 

I added this to my function:
 function  afterDataSetFilled( series, dataSet, icsc ) {
	Packages.java.lang.System.out.println("we good? " + row["typeCol"]);
 }

and got: 
org.eclipse.birt.chart.exception.ChartException: TypeError: Cannot read property "typeCol" from undefined at line 35 of chart script:''

Which means row is undefined (or null) inside the function. Maybe I need to be inside a different function - I don't know.

**Reply by hvbtup** [reply]
Of course it is undefined. You have to supply `row` as an argument to your function when you call it from the event script.

**Reply by speckyspooky** [reply]
The chart methods are a kind of own family.
Have you tried your variable "dataSet" like dataSet["typeCol"]?

**Reply by arass** [reply]
In BIRT scripting, the afterDataSetFilled function is a predefined event handler, and the parameters it accepts (series, dataSet, icsc) are fixed. @hvbtup 

yeah, dataSet just has the 2-3 values that got grouped to do this graph. There's nothing there. 
Method gets called separately for X and Y axis. With just the values that go into that. @speckyspooky 

Speckyspooky is correct "own family". Bunch of stuff doesn't work the same inside. Can't even reference customer's Java packages directly. It's weird they chose to go this way. Horrifically inconsistent with the rest of the scripting.

**Comment by speckyspooky** [comment]
Additional hint, I don't know exactly your use case.
In my case I created a maximum value from a dataset "onFetch()".
Saved the value as persistent variable and used it at chart-side to set the maxmimum value for correct scaling.

Here is my case as screen - onFetch() (DataSet) & onRender() (Chart):

![grafik](https://github.com/user-attachments/assets/6fa733be-d087-4ded-900d-dccda0034d97)

**Reply by arass** [reply]
It's a "running report", so I can't do a global Max. Thanks. Otherwise this would have been great. 

The data is grouped and for each group it creates a page with several charts and graphs. 
fetch iterates over:
row[1] = date
row[2] = chartType
row[3] = val4chart

So when I look at dataSet - it's just an array of val4charts.
And if these values are chartType of "bolts" - I need to setScale().

It's far more complicated, but that's the gist. :) 

I just wanna be able to get to my row[2] for this group that's being charted. 
Since the section (label) name is also  listed above on the page - I thought this would be easier.

**Reply by speckyspooky** [reply]
You can use the global persist variables to create an associated array with your information based as key-value-pair
and use this array to get the information at chart level.

**Reply by arass** [reply]
I don't have chartType inside the script (so far). 
I only have an array of val4chart
And val4chart values are not unique in any way. So I can't do a lookup from val4chart to chartType. 
So I don't know/see how I can get my chartType.

**Comment by hvbtup** [comment]
> You can use the global persist variables to create an associated array with your information based as key-value-pair
and use this array to get the information at chart level.

And you can fill this global variable e..g in the `onFetch` event of the data set.

Another trick that is sometimes useful is to use the same dataset twice.
First, you use it for an list item which is hidden, the whole purpose of this list item is to fill some global variables.

Then you can use these variables later (further down) in the report layout for other purposes.

**Reply by speckyspooky** [reply]
...and additional option:
When you have your charts in a table then you can store the value to a global variable at `onCreate` **of the row** and so you would be able to use this for your chart.

**Comment by arass** [comment]
onCreate? where would that be? I don't see it in any dropdowns.

**Reply by hvbtup** [reply]
Please read Thomas` answer carefully. He precisely wrote where to set the global var.

**Reply by speckyspooky** [reply]
The topic is that we didn't know the structure of your report.
If you have one chart which is part of a DataTable then you have the "row"-object and there you have the onCreate() method.

If you have 10 single charts then you have to use another way.
The simplste way could be that you make your entries of your chartTypes uniquie may be based on a associated array with key&value
or you create a multidimensional array with your own key to handle the type records to get your chartType information
